### ISSUE TITLE
Lua module 'redis' not found. No field package.preload['redis']

### ISSUE THREAD

**Issue by sivaraman-27** [haproxy/haproxy#2421]
### Detailed Description of the Problem

I have a used lua script to dynamically select the backend. In that script, tried to use redis (https://github.com/nrk/redis-lua). However, when a client sends a request, this lua script is successfully called but Haproxy can't somehow load the redis module

**haproxy.cfg** (Simplified contents):
```
global
  lua-load /etc/haproxy/select_backend.lua
  log         127.0.0.1 local2
  
  chroot      /var/lib/haproxy
  pidfile     /var/run/haproxy.pid
  maxconn     4000
  user        haproxy
  group       haproxy
  daemon

  stats socket /var/lib/haproxy/stats
  ssl-default-bind-ciphers PROFILE=SYSTEM
  ssl-default-server-ciphers PROFILE=SYSTEM

defaults
  mode                    http
  log                     global
  option                  httplog
  option                  dontlognull
  option http-server-close
  option forwardfor       except 127.0.0.0/8
  option                  redispatch
 
frontend testFrnt
  bind *:80
  use_backend %[lua.select_backend]

backend testBck
  server testBck1 myserver.example.com:443 check
```

**select_backend.lua**

```
function select_backend(txn)
    package.path  = package.path  .. ";/redisd/?.lua"
    core.log(core.debug, "Path: " .. package.path)
    core.log(core.debug, "CPath: " .. package.cpath)
    local redis = require "redis"
 end
  core.register_fetches("select_backend", select_backend)
```

**haproxy.log** (showing no file '/redisd/redis.lua'):
```
Jan 23 10:04:05 localhost haproxy[416086]: Path: /usr/share/lua/5.4/?.lua;/usr/share/lua/5.4/?/init.lua;/usr/lib64/lua/5.4/?.lua;/usr/lib64/lua/5.4/?/init.lua;./?.lua;./?/init.lua;/redisd/redis.lua
Jan 23 10:04:05 localhost haproxy[416086]: CPath: /usr/lib64/lua/5.4/?.so;/usr/lib64/lua/5.4/loadall.so;./?.so
Jan 23 10:04:05 localhost haproxy[416086]: Lua sample-fetch 'select_backend': [state-id 0] runtime error: /etc/haproxy/select_backend.lua:5: module 'redis' not found:
    #012#011no field package.preload['redis']
    #012#011no file '/usr/share/lua/5.4/redis.lua'
    #012#011no file '/usr/share/lua/5.4/redis/init.lua'
    #012#011no file '/usr/lib64/lua/5.4/redis.lua'
    #012#011no file '/usr/lib64/lua/5.4/redis/init.lua'
    #012#011no file './redis.lua'
    #012#011no file './redis/init.lua'
    🔴🔴🔴🔴🔴🔴🔴 #012#011no file '/redisd/redis.lua' 🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴
    #012#011no file '/usr/lib64/lua/5.4/redis.so'
    #012#011no file '/usr/lib64/lua/5.4/loadall.so'
    #012#011no file './redis.so' from [C]: in global 'require', /etc/haproxy/select_backend.lua:5: in function line 1.
```

 **/redisd/redis.lua** is present
```
[root@ip-xxx /]# ls -ld /redisd/
                 drwxrwxrwx. 2 root root 23 Jan 23 07:35 /redisd/
[root@ip-xxx /]# ls -l /redisd/redis.lua
                 -rwxrwxrwx. 1 root root 35599 Jan 22 16:14 /redisd/redis.lua
```

**Note**
This issue only occurs inside Haproxy lua execution. If I a create a sample lua file with below contents and execute, then there is no problem
```
package.path = package.path  .. ";/redisd/?.lua"
local redis = require "redis"
```

### Expected Behavior

Haproxy should load redis module without any error

### Steps to Reproduce the Behavior

--

### What is your configuration?

```haproxy
--
```

### Output of `haproxy -vv`

```plain
Haproxy:
HAProxy version 2.8.0-fdd8154 2023/05/31 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.0.html
Running on: Linux 6.1.xx.amzn2023.x86_64 #1 SMP PREEMPT_DYNAMIC Thu Jun 29 18:11:59 UTC 2023 x86_64

Lua:
Lua 5.4.4  Copyright (C) 1994-2022 Lua.org, PUC-Rio
```

### Additional Information

1. I have tried giving access to all directories/subdirectories/required lua module file like mentioned in the below similar github issues (https://github.com/openresty/lua-resty-redis/issues/24, https://github.com/openresty/lua-nginx-module/issues/792)
2. I have also moved the redis.lua file to /tmp/ folder

I'm still facing this issue

**Comment by sivaraman-27** [comment]
@lukastribus @Tristan971 Please help resolve this issue. For some reason, Haproxy lua environment is not able recognize the redis module present (eventhough the file path is correct)

**Comment by Tristan971** [comment]
Wrong place to ask (should be mailing list, Discourse, or Slack). Also don't ping random people.

But since it's there, and in case someone else searches for it, here's what I got to work. I'm **not** an expert on the topic and this might not be the best way.

- First: I manage LUA packages using LuaRocks (it seemed like the only sane option to me).
- Second: In HAProxy's global config:
```
global
  ...
  lua-prepend-path {{ lua_path }} path
  lua-prepend-path {{ lua_cpath }} cpath

...
```

with:
- `{{ lua_path }}` taken from `luarocks --lua-version {{ haproxy_lua_version }} path --lr-path`
- `{{ lua_cpath }}` taken from `luarocks --lua-version {{ haproxy_lua_version }} path --lr-cpath`

Which then allows my `require`s to work.

You probably have to make sure that the LUA versions HAProxy and LuaRocks use are matching.
Or maybe not. I don't know.
Best option is to have only 1 version installed wich matches the one HAProxy is compiled against (see `haproxy -vv`). Then you can't get it wrong.

**Comment by sivaraman-27** [comment]
Hi @Tristan971,
**I certainly think this is the only place for this issue** because the file '/redisd/redis.lua' is present and running a lua script outside haproxy environment loads this file and executes without any exception

But inside haproxy environment, lua is able to load the file path (see below) but saying it's not that file.
```
Jan 23 10:04:05 localhost haproxy[416086]: Lua sample-fetch 'select_backend': [state-id 0] runtime error: /etc/haproxy/select_backend.lua:5: module 'redis' not found:
    #012#011no file './redis.lua'
    #012#011no file './redis/init.lua'
    🔴🔴🔴🔴🔴🔴🔴 #012#011no file '/redisd/redis.lua' 🔴🔴🔴🔴🔴🔴🔴🔴🔴🔴
    #012#011no file '/usr/lib64/lua/5.4/redis.so'
```

This is happening because clearly there is some bug in your haproxy - lua integration




_FYI_: I have tried your answer, instead of doing _package.path  = package.path  .. ";/redisd/?.lua"_, you've used  lua-prepend-path. It shows the same error (it sees the file path, but saying it's not that file)

**Comment by Darlelet** [comment]
No, in fact it is expected behavior.

You are performing the require very late (ie: during runtime), and your haproxy is configured to execute itself within a chroot.
``` 
chroot      /var/lib/haproxy
```
(taken from your config snippet)

Thus, once haproxy is started, haproxy will be confined to `/var/lib/haproxy` parent directory, and as the embedded lua engine is part of haproxy process, it makes no exception to that.

Either you need to perform the require early during the initial lua-load, that is during the init step, which occurs before the chroot (by performing the `require` at the top of the lua script, or in a `register_init()` routine, but not within a function that will be executed during runtime), or you need to adjust your environment or configuration so that the file is accessible from the chrooted location.

**Comment by sivaraman-27** [comment]
@Darlelet Thank you so much. I failed to notice that line. Now, everything works great

@Tristan971 , The chroot was already present when I installed Haproxy. 
Why chroot is needed in Haproxy?
Can I just remove that?
Any performace hit will be there?

**Comment by Tristan971** [comment]
All your questions are answered [here](https://docs.haproxy.org/2.9/configuration.html).

Now please stop pinging me.

**Comment by sivaraman-27** [comment]
Hi,
Apologies for replying to this closed thread. I've encountered an issue which is related to this issue.
I installed haproxy using 'yum' package manager which also installed lua.
Package.path for this lua:
`/usr/share/lua/5.4/?.lua;/usr/share/lua/5.4/?/init.lua;/usr/lib64/lua/5.4/?.lua;/usr/lib64/lua/5.4/?/init.lua;./?.lua;./?/init.lua`

To use external packages in haproxy-lua script (e.g: redis-lua), we need to install them with '[luarocks](https://luarocks.org/)' (this is the preferred option for installing lua packages). While installing the luarocks, faced with the below error

```
$ ./configure && make && sudo make install

Configuring LuaRocks version 3.9.2...

Lua interpreter not found in /usr/share/lua/5.4/bin
You may want to use the flags --with-lua, --with-lua-bin and/or --lua-version
Run ./configure --help for details.

configure failed.
```


In the above, luarocks couldn't find the haproxy-lua. So, I installed another lua (not related to haproxy) and this has a different 'package.path' than the haproxy-lua. Then installed luarocks (here luarocks detects the external lua which is not associated with haproxy).
If a install a package, it will be present in the external lua package.path but not in haproxy-lua path. So, essentially if want to extend haproxy with lua, I have to manually modify the haproxy-lua like below
`package.path of haproxy-lua = package.path of haproxy-lua + package.path of external lua `

**Is there any other way around this (want to install luarocks pointing to haproxy-lua itself. Don't want to modify the package.path of haproxy-lua).**

**Note:** (The below method fixes the issue, but I want to avoid the below method)
If I install lua first and build haproxy from source code, then haproxy takes the existing lua path. So, I can now install luarocks pointing to haproxy-lua (or external lua) as both lua are same.
